# Taller 5 - TF - IDF


### Andrés Felipe Murillo

## Punto 1

### Pre-Procesamiento

In [83]:
import re
import pandas as pd
from collections import Counter

from nltk.corpus import stopwords
stopwords_sp = stopwords.words('spanish')

from nltk.stem.snowball import SnowballStemmer
spanishStemmer = SnowballStemmer("spanish")

from sklearn.feature_extraction.text import CountVectorizer

In [84]:
bob_esponja = pd.read_csv("C:/Users/AndresFelipe/Downloads/bob_esponja (1).csv")
bob_esponja

Personaje  \
0  Bob Esponja Pantalones Cuadrados   
1                 Patricio Estrella   
2              Calamardo Tentáculos   
3                  Arenita Mejillas   
4               Eugenio H. Cangrejo   
5               Sheldon J. Plankton   
6                    Karen Plankton   
7                  Perlita Cangrejo   
8                         Sra. Puff   
9                   Gary el Caracol   

                                         Descripción  
0  Bob Esponja es una esponja de mar con forma re...  
1  Es una estrella de mar de color rosa, el mejor...  
2  Calamardo Tentáculos es un personaje principal...  
3  Arenita Mejillas (Sandy Cheeks en inglés) en u...  
4  Don Eugenio H. Cangrejo (normalmente llamado S...  
5  Sheldon J. Plankton, o simplemente Plankton, e...  
6  Karen Plankton es uno de los dos principales a...  
7  Perla "Perlita" Cangrejo es un personaje princ...  
8  La Señora Puff es un personaje principal de Bo...  
9  Gary el Caracol es un personaje principal de B...

In [92]:
def pre_procesado(texto):
    texto = re.sub(r"[\W\d_]+", " ", texto.lower())
    texto = [palabra for palabra in texto.split() if palabra not in stopwords_sp]
    texto = " ".join(texto)
    return texto

bob_esponja['pre-procesado'] = bob_esponja['Descripción'].apply(lambda texto: pre_procesado(texto))
bob_esponja

Personaje  \
0  Bob Esponja Pantalones Cuadrados   
1                 Patricio Estrella   
2              Calamardo Tentáculos   
3                  Arenita Mejillas   
4               Eugenio H. Cangrejo   
5               Sheldon J. Plankton   
6                    Karen Plankton   
7                  Perlita Cangrejo   
8                         Sra. Puff   
9                   Gary el Caracol   

                                         Descripción  \
0  Bob Esponja es una esponja de mar con forma re...   
1  Es una estrella de mar de color rosa, el mejor...   
2  Calamardo Tentáculos es un personaje principal...   
3  Arenita Mejillas (Sandy Cheeks en inglés) en u...   
4  Don Eugenio H. Cangrejo (normalmente llamado S...   
5  Sheldon J. Plankton, o simplemente Plankton, e...   
6  Karen Plankton es uno de los dos principales a...   
7  Perla "Perlita" Cangrejo es un personaje princ...   
8  La Señora Puff es un personaje principal de Bo...   
9  Gary el Caracol es un personaje principal de B...   

                                       pre-procesado  
0  bob esponja esponja mar forma rectangular colo...  
1  estrella mar color rosa mejor amigo bob esponj...  
2  calamardo tentáculos personaje principal bob e...  
3  arenita mejillas sandy cheeks inglés personaje...  
4  don eugenio h cangrejo normalmente llamado señ...  
5  sheldon j plankton simplemente plankton dos an...  
6  karen plankton dos principales antagonistas bo...  
7  perla perlita cangrejo personaje principal bob...  
8  señora puff personaje principal bob esponja ma...  
9  gary caracol personaje principal bob esponja q...

In [98]:
# Stemming
from nltk.stem.porter import PorterStemmer

def get_stemmed_text(corpus):
    stemmer = PorterStemmer()
    return [' '.join([stemmer.stem(word) for word in review.split()]) for review in corpus]

bob_esponja['stemmedtext'] = get_stemmed_text(bob_esponja['pre-procesado'])
bob_esponja

Personaje  \
0  Bob Esponja Pantalones Cuadrados   
1                 Patricio Estrella   
2              Calamardo Tentáculos   
3                  Arenita Mejillas   
4               Eugenio H. Cangrejo   
5               Sheldon J. Plankton   
6                    Karen Plankton   
7                  Perlita Cangrejo   
8                         Sra. Puff   
9                   Gary el Caracol   

                                         Descripción  \
0  Bob Esponja es una esponja de mar con forma re...   
1  Es una estrella de mar de color rosa, el mejor...   
2  Calamardo Tentáculos es un personaje principal...   
3  Arenita Mejillas (Sandy Cheeks en inglés) en u...   
4  Don Eugenio H. Cangrejo (normalmente llamado S...   
5  Sheldon J. Plankton, o simplemente Plankton, e...   
6  Karen Plankton es uno de los dos principales a...   
7  Perla "Perlita" Cangrejo es un personaje princ...   
8  La Señora Puff es un personaje principal de Bo...   
9  Gary el Caracol es un personaje principal de B...   

                                       pre-procesado  \
0  bob esponja esponja mar forma rectangular colo...   
1  estrella mar color rosa mejor amigo bob esponj...   
2  calamardo tentáculos personaje principal bob e...   
3  arenita mejillas sandy cheeks inglés personaje...   
4  don eugenio h cangrejo normalmente llamado señ...   
5  sheldon j plankton simplemente plankton dos an...   
6  karen plankton dos principales antagonistas bo...   
7  perla perlita cangrejo personaje principal bob...   
8  señora puff personaje principal bob esponja ma...   
9  gary caracol personaje principal bob esponja q...   

                                         stemmedtext  
0  bob esponja esponja mar forma rectangular colo...  
1  estrella mar color rosa mejor amigo bob esponj...  
2  calamardo tentáculo personaj princip bob espon...  
3  arenita mejilla sandi cheek inglé personaj pri...  
4  don eugenio h cangrejo normalment llamado seño...  
5  sheldon j plankton simplement plankton do anta...  
6  karen plankton do principal antagonista bob es...  
7  perla perlita cangrejo personaj princip bob es...  
8  señora puff personaj princip bob esponja maest...  
9  gari caracol personaj princip bob esponja quer...

## Punto 2 

### Matriz TF - IDF

In [106]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer()
tfidf = tfidf_vect.fit_transform(bob_esponja['stemmedtext'].values)

tfidf_matrix = pd.DataFrame(tfidf.toarray(), columns=tfidf_vect.get_feature_names())
tfidf_matrix.index = bob_esponja.index

tfidf_matrix.T.round(3)

0    1      2      3      4    5      6    7      8      9
abajo      0.057  0.0  0.000  0.000  0.000  0.0  0.000  0.0  0.000  0.000
aborrec    0.000  0.0  0.091  0.000  0.000  0.0  0.000  0.0  0.000  0.000
acaba      0.000  0.0  0.091  0.000  0.000  0.0  0.000  0.0  0.000  0.000
accident   0.048  0.0  0.000  0.000  0.045  0.0  0.000  0.0  0.000  0.000
aceptó     0.000  0.0  0.000  0.000  0.000  0.0  0.064  0.0  0.000  0.000
...          ...  ...    ...    ...    ...  ...    ...  ...    ...    ...
zapato     0.042  0.0  0.000  0.000  0.000  0.0  0.000  0.0  0.093  0.068
árbol      0.000  0.0  0.000  0.081  0.000  0.0  0.000  0.0  0.000  0.000
éste       0.000  0.0  0.000  0.000  0.000  0.0  0.000  0.0  0.000  0.091
única      0.057  0.0  0.000  0.000  0.000  0.0  0.000  0.0  0.000  0.000
únicament  0.000  0.0  0.000  0.000  0.053  0.0  0.000  0.0  0.000  0.000

[799 rows x 10 columns]

## Punto 3

### Distancia del Coseno

In [104]:
from sklearn.metrics.pairwise import cosine_distances

dist_cos = cosine_distances(tfidf_matrix.T.values)
dist_cos = pd.DataFrame(dist_cos, columns = tfidf_matrix.columns, index = tfidf_matrix.columns)
dist_cos

0         1         2         3         4         5         6  \
0  0.000000  0.850507  0.927257  0.871921  0.881220  0.901743  0.957920   
1  0.850507  0.000000  0.808538  0.893063  0.893730  0.968058  0.961623   
2  0.927257  0.808538  0.000000  0.943984  0.873121  0.921201  0.969450   
3  0.871921  0.893063  0.943984  0.000000  0.927035  0.982037  0.961451   
4  0.881220  0.893730  0.873121  0.927035  0.000000  0.704685  0.903820   
5  0.901743  0.968058  0.921201  0.982037  0.704685  0.000000  0.808370   
6  0.957920  0.961623  0.969450  0.961451  0.903820  0.808370  0.000000   
7  0.916038  0.936298  0.972910  0.982508  0.829929  0.921811  0.984115   
8  0.882641  0.954744  0.968216  0.983039  0.938435  0.990016  0.991454   
9  0.931071  0.922686  0.914684  0.930316  0.928366  0.971474  0.978322   

          7         8         9  
0  0.916038  0.882641  0.931071  
1  0.936298  0.954744  0.922686  
2  0.972910  0.968216  0.914684  
3  0.982508  0.983039  0.930316  
4  0.829929  0.938435  0.928366  
5  0.921811  0.990016  0.971474  
6  0.984115  0.991454  0.978322  
7  0.000000  0.895400  0.967919  
8  0.895400  0.000000  0.963591  
9  0.967919  0.963591  0.000000

## ¿Cuáles son los personajes más parecidos?

Los personajes mas parecidos son Karen Plankton(No.6) y la Sra. Puff(No.8) con una metrica de 0.991454

## ¿Cuáles son los personajes más diferentes?

Los personajes mas diferentes son Sheldon J. Plankton(No.5) y la Eugenio H. Cangrejo(No.4) con una metrica de 0.704685